File: app/src/main/java/com/arr/angel/pertpratice/db/TopicRepository.java
package com.arr.angel.pertpratice.db;

import android.app.Application;
import android.arch.lifecycle.LiveData;

import com.arr.angel.pertpratice.model.Topic;
import com.arr.angel.pertpratice.util.TopicData;

import java.util.List;
import java.util.concurrent.Executor;

public class TopicRepository {


    private final TopicDao mTopicDao;
    private LiveData<List<Topic>> mTopicsListLiveData;
    private LiveData<Topic> mTopicLiveData;
    private List<Topic> topics;
    private final Executor executor;

    public TopicRepository(Application application, Executor executor) {
        AppDatabase db = AppDatabase.getInstance(application);
        mTopicDao = db.mTopicDao();
        this.executor = executor;
    }

    public LiveData<List<Topic>> getTopicsList() {
        mTopicsListLiveData = mTopicDao.loadAllTopics();
        return mTopicsListLiveData;
    }

    public void deleteTopic(final Topic topic) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                mTopicDao.deleteTopic(topic);
            }
        });
    }

    public void insertTopic(final Topic topic) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                mTopicDao.insertTopic(topic);
            }
        });
    }

    public void updateTopic(final Topic topic) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                mTopicDao.updateTopic(topic);
            }
        });
    }

    public LiveData<Topic> getTopicFromDB(int topicId) {
        mTopicLiveData = mTopicDao.loadTopicById(topicId);
        return mTopicLiveData;
    }

    private void addTopicsToDB() {
        topics = new TopicData().getTopicList();
        executor.execute(new Runnable() {
            @Override
            public void run() {
                for (Topic t : topics) {
                    mTopicDao.insertTopic(t);
                }
            }
        });
    }

    public List<Topic> getTopics() {
        return topics;
    }

    public void loadData(){
        addTopicsToDB();
    }
}

File: app/src/main/java/com/arr/angel/pertpratice/db/TopicRepository.kt
package com.arr.angel.pertpratice.db

import android.content.Context
import androidx.lifecycle.LiveData
import com.arr.angel.pertpratice.model.Topic
import com.arr.angel.pertpratice.util.TopicData
import kotlinx.coroutines.launch

class TopicRepository(ctx: Context) {
    val topics : List<Topic>? = TopicData().getTopicList()
    lateinit var mTopicsListLiveData : LiveData<List<Topic>>
    lateinit var mTopicLiveData : LiveData<Topic>

    private val topicDao by lazy { AppDatabase.getInstance(ctx).mTopicDao() }

    fun getTopicsList() : LiveData<List<Topic>> {
        mTopicsListLiveData = topicDao.loadAllTopics()
        return mTopicsListLiveData
    }

    fun getTopicFromDB(topicId : Int) : LiveData<Topic> {
        mTopicLiveData = topicDao.loadTopicById(topicId)
        return mTopicLiveData
    }

    fun addTopicsToDB() {
        for (topic in topics?: emptyList()) {
            dbScope.launch {
                topicDao.insertTopic(topic)
            }
        }
    }

    fun updateTopic(topic: Topic) {
        dbScope.launch { topicDao.updateTopic(topic) }
    }

    fun deleteTopic(topic: Topic) {
        dbScope.launch { topicDao.deleteTopic(topic) }
    }

    fun insertTopic(topic: Topic) {
        dbScope.launch { topicDao.insertTopic(topic) }
    }

    fun loadData() {
        addTopicsToDB()
    }
}